File: src/main/java/fr/lewon/bot/runner/lifecycle/task/TaskLifeCycleOperation.java
package fr.lewon.bot.runner.lifecycle.task;

import fr.lewon.bot.runner.lifecycle.ILifeCycleOperation;
import fr.lewon.bot.runner.lifecycle.Transition;

public enum TaskLifeCycleOperation implements ILifeCycleOperation<TaskState> {

    START(new Transition<>(TaskState.ACTIVE, TaskState.PENDING)),
    STOP(new Transition<>(TaskState.DISPOSED, TaskState.ACTIVE));

    private final Transition<TaskState>[] transitions;

    TaskLifeCycleOperation(Transition<TaskState>... transitions) {
        this.transitions = transitions;
    }

    @Override
    public Transition<TaskState>[] getTransitions() {
        return this.transitions;
    }
}

File: src/main/kotlin/fr/lewon/bot/runner/lifecycle/task/TaskLifeCycleOperation.kt
package fr.lewon.bot.runner.lifecycle.task

import fr.lewon.bot.runner.lifecycle.ILifeCycleOperation

enum class TaskLifeCycleOperation constructor(private val to: TaskState, private vararg val from: TaskState) : ILifeCycleOperation<TaskState> {

    START(TaskState.ACTIVE, TaskState.PENDING),
    STOP(TaskState.DISPOSED, TaskState.ACTIVE),
    CRASH(TaskState.CRASHED);

    override fun getTo(): TaskState {
        return to
    }

    override fun getFrom(): List<TaskState> {
        return listOf(*from)
    }

}
